### PR TITLE
Extract method params and return types from jsdoc

### LINKED ISSUE #585: Extract method params and return types from jsdoc

### PR DESCRIPTION
**Pull request by aomarks**
Fixes #585 

 - [x] CHANGELOG.md has been updated

### CODE REVIEW

File: src/polymer/js-utils.ts [hunk 1]
```diff
@@ -101,10 +101,38 @@ export function toScannedMethod(
...
-      // formatting to escodegen.
-      return {name: escodegen.generate(param)};
+
+    const paramTags = {};
+    if (scannedMethod.jsdoc) {
+      for (const tag of (scannedMethod.jsdoc.tags || [])) {
+        if (tag.tag === 'param' && tag.name) {
+          paramTags[tag.name] = tag;
+
+        } else if (tag.tag === 'returns') {
```
Comment (azakus) [inline src/polymer/js-utils.ts]: According to https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System#typedefs, `@return` is the correct tag, though both work.

Comment (rictic) [inline src/polymer/js-utils.ts]: +1 to also matching `@return`

**Comment by azakus** [comment]
Hmm, there's some oddities in the reified closure types
Somehow `@param {string | !Array<number|string>} path` is turning into
`@param {(string|!Array.<(number|string)>)} path` which is invalid syntax

File: src/polymer/js-utils.ts [hunk 2]
```diff
@@ -101,10 +101,38 @@ export function toScannedMethod(
...
+            type: tag.type || undefined,
+            desc: tag.description || undefined,
+          };
+        }
+      }
+    }
+
+    scannedMethod.params = (value.params || []).map((nodeParam) => {
+      const param: MethodParam = {
+        // With ES6 we can have a lot of param patterns.Best to leave the
```
Comment (rictic) [inline src/polymer/js-utils.ts]: space after the `.` in this comment plz

**Comment by azakus** [comment]
@justinfagnani points out that the `@param {(string|!Array.<(number|string)>)}` may actually be valid in closure, so I'm giving it a shot with a full compile pass

**Comment by rictic** [comment]
@azakus `Array.<string>` was valid closure at one point, they're trying to move everyone to `Array<string>` though. Where is the dot coming from? Is the jsdoc node library adding it or is it in the source?

File: src/polymer/js-utils.ts (hunk 2 above)
Comment (aomarks) [inline src/polymer/js-utils.ts]: Done.

File: src/polymer/js-utils.ts (hunk 1 above)
Comment (aomarks) [inline src/polymer/js-utils.ts]: Done.

Funny that http://usejsdoc.org/tags-returns.html likes `@returns` as canonical.

**Comment by azakus** [comment]
@rictic nope, certainly not in the source: https://github.com/Polymer/polymer/blob/master/lib/mixins/property-effects.html#L1391

**Pull request merged**